variable is assigned to the wrong type

Hello,

sorry for the vague title, I don't know how to describe this concisely.

The following code:
```
const std = @import("std");

const BindMarker = union(enum) {
    Type: type,
    NoType: void,
};

const ParsedQuery = struct {
    const Self = @This();

    bind_markers: [128]BindMarker,
    nb_bind_markers: usize,

    fn from(comptime n: comptime_int) Self {
        comptime var parsed_query: ParsedQuery = undefined;

        if (n == 0) {
            parsed_query.bind_markers[0] = BindMarker{ .Type = usize };
            parsed_query.bind_markers[1] = BindMarker{ .Type = usize };
        } else {
            parsed_query.bind_markers[0] = BindMarker{ .Type = usize };
            parsed_query.bind_markers[1] = BindMarker{ .Type = u32 };
        }

        parsed_query.nb_bind_markers = 2;

        return parsed_query;
    }

    fn getString(comptime self: *Self) []const u8 {
        return self.str[0..self.str_len];
    }
};

fn Statement(comptime query: ParsedQuery) type {
    return struct {
        const Self = @This();

        pub fn prepare() Self {
            return Self{};
        }

        pub fn bind(self: *Self, values: anytype) void {
            const StructType = @TypeOf(values);
            const StructTypeInfo = @typeInfo(StructType).Struct;

            if (comptime query.nb_bind_markers != StructTypeInfo.fields.len) {
                @compileError("number of bind markers not equal to number of fields");
            }

            inline for (StructTypeInfo.fields) |struct_field, _i| {
                if (struct_field.field_type != query.bind_markers[_i].Type) {
                    @compileError("value type " ++ @typeName(struct_field.field_type) ++ " is not the bind marker type " ++ @typeName(query.bind_markers[_i].Type));
                }
            }
        }
    };
}

pub fn main() anyerror!void {
    comptime var parsed_query1 = ParsedQuery.from(0);
    comptime var parsed_query2 = ParsedQuery.from(1);

    var stmt1 = Statement(parsed_query1).prepare();
    stmt1.bind(.{
        .a = @as(usize, 0),
        .b = @as(usize, 0),
    });

    var stmt2 = Statement(parsed_query2).prepare();
    stmt2.bind(.{
        .a = @as(usize, 0),
        .b = @as(u32, 0),
    });
}
```

fails to compile with the following error:
```
./src/main.zig:53:21: error: value type u32 is not the bind marker type usize
                    @compileError("value type " ++ @typeName(struct_field.field_type) ++ " is not the bind marker type " ++ @typeName(query.bind_markers[_i].Type));
                    ^
./src/main.zig:71:15: note: called from here
    stmt2.bind(.{
              ^
./src/main.zig:60:29: note: called from here
pub fn main() anyerror!void {
```

it looks like `stmt2` acts as if it is the type `Statement(parsed_query1)` when it clearly should be `Statement(parsed_query2)`.

--- a/src/stage1/analyze.cpp
+++ b/src/stage1/analyze.cpp
@@ -7079,8 +7079,6 @@ bool const_values_equal(CodeGen *g, ZigValue *a, ZigValue *b) {
             if (bigint_cmp(&union1->tag, &union2->tag) == CmpEQ) {
                 TypeUnionField *field = find_union_field_by_tag(a->type, &union1->tag);
                 assert(field != nullptr);
-                if (!type_has_bits(g, field->type_entry))
-                    return true;
                 assert(find_union_field_by_tag(a->type, &union2->tag) != nullptr);
                 return const_values_equal(g, union1->payload, union2->payload);
             }

--- a/test/stage1/behavior.zig
+++ b/test/stage1/behavior.zig
@@ -56,6 +56,7 @@ comptime {
     _ = @import("behavior/bugs/6456.zig");
     _ = @import("behavior/bugs/6781.zig");
     _ = @import("behavior/bugs/6850.zig");
+    _ = @import("behavior/bugs/7047.zig");
     _ = @import("behavior/bugs/394.zig");
     _ = @import("behavior/bugs/421.zig");
     _ = @import("behavior/bugs/529.zig");

--- a/test/stage1/behavior/bugs/7047.zig
+++ b/test/stage1/behavior/bugs/7047.zig
@@ -1,0 +1,22 @@
+const std = @import("std");
+
+const U = union(enum) {
+    T: type,
+    N: void,
+};
+
+fn S(comptime query: U) type {
+    return struct {
+        fn tag() type {
+            return query.T;
+        }
+    };
+}
+
+test "compiler doesn't consider equal unions with different 'type' payload" {
+    const s1 = S(U{ .T = u32 }).tag();
+    std.testing.expectEqual(u32, s1);
+
+    const s2 = S(U{ .T = u64 }).tag();
+    std.testing.expectEqual(u64, s2);
+}

